### PR TITLE
feat: move to scala 2.13, spark 3.2.1 and fix spark dependencies scope

### PR DESCRIPTION
**Pull request by dohr-michael**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -52,7 +52,7 @@ repositories {
 extra.apply {
     set("annotationsVersion", "22.0.0")
     set("mongodbDriverVersion", "[4.5.0,4.5.99)")
-    set("sparkVersion", "3.1.2")
+    set("sparkVersion", "3.2.2")
```
Comment (cchantep) [inline build.gradle.kts:55]: Would allow to set it using a sys prop, to be able to cross build.

Comment (cchantep) [reply build.gradle.kts:55]: Would also add a similar `scalaBinaryVersion` to allow to cross build between 2.1X.
This should also be use as suffix for the artifactId in publication, as this lib in not a pure Java one, but a Scala one depending on Scala library version which can be cross built accordingly.

**Review by cchantep — CHANGES_REQUESTED** [review]

File: build.gradle.kts (hunk 1 above)
Comment (dohr-michael) [reply build.gradle.kts:55]: Yea,
I've already mention this point in another pr (https://github.com/mongodb/mongo-spark/pull/58#issuecomment-1219430851).
To have cross build, we need to have specifics sources for 2.12 and 2.13 (implicits conversion from java to scala in src/main/java/com/mongodb/spark/sql/connector/schema/BsonDocumentToRowConverter.java) and i'm not a gradle power user to do that :D, 
i'm more comfortable with sbt do to that

File: build.gradle.kts [hunk 2]
```diff
@@ -52,7 +52,7 @@ repositories {
 extra.apply {
```
Comment (cchantep) [inline build.gradle.kts:52]:
```suggestion
// Usage: gradle -DscalaBinaryVersion=2.13 -DsparkVersion=3.2.2
def scalaBinaryVersion = findProperty("scalaBinaryVersion") ?: "2.12"
def sparkVersion = findProperty("sparkVersion") ?: "3.1.2"

extra.apply {
```

File: build.gradle.kts (hunk 1 above)
Comment (cchantep) [inline build.gradle.kts:55]:
```suggestion

    set("scalaBinaryVersion", scalaBinaryVersion)
    set("sparkVersion", sparkVersion)
```

File: build.gradle.kts [hunk 3]
```diff
@@ -66,10 +66,14 @@ extra.apply {
...
 
-    implementation("org.apache.spark:spark-core_2.12:${project.extra["sparkVersion"]}")
-    implementation("org.apache.spark:spark-sql_2.12:${project.extra["sparkVersion"]}")
-    implementation("org.apache.spark:spark-catalyst_2.12:${project.extra["sparkVersion"]}")
-    implementation("org.apache.spark:spark-streaming_2.12:${project.extra["sparkVersion"]}")
+    // dohr-michael : this dependencies must be 'provided scope' because in all spark project this dependencies
+    // are already bundle to spark runtime, if we use implementation, this deps will be include in the project
+    // and by this way loaded 2 times in the container / final bundle of spark project.
+    // We need to include this deps to testImplementation because gradle don't include it in any runtime.
+    compileOnly("org.apache.spark:spark-core_2.13:${project.extra["sparkVersion"]}")
```
Comment (cchantep) [inline build.gradle.kts:73]:
```suggestion
    compileOnly("org.apache.spark:spark-core_${project.extra["scalaBinaryVersion"]}:${project.extra["sparkVersion"]}")
```

File: build.gradle.kts [hunk 4]
```diff
@@ -66,10 +66,14 @@ extra.apply {
...
+    compileOnly("org.apache.spark:spark-sql_2.13:${project.extra["sparkVersion"]}")
```
Comment (cchantep) [inline build.gradle.kts:74]:
```suggestion
    compileOnly("org.apache.spark:spark-sql_${project.extra["scalaBinaryVersion"]}:${project.extra["sparkVersion"]}")
```

File: build.gradle.kts [hunk 5]
```diff
@@ -66,10 +66,14 @@ extra.apply {
...
+    compileOnly("org.apache.spark:spark-catalyst_2.13:${project.extra["sparkVersion"]}")
```
Comment (cchantep) [inline build.gradle.kts:75]:
```suggestion
    compileOnly("org.apache.spark:spark-catalyst_${project.extra["scalaBinaryVersion"]}:${project.extra["sparkVersion"]}")
```

File: build.gradle.kts [hunk 6]
```diff
@@ -66,10 +66,14 @@ extra.apply {
...
+    compileOnly("org.apache.spark:spark-streaming_2.13:${project.extra["sparkVersion"]}")
```
Comment (cchantep) [inline build.gradle.kts:76]:
```suggestion
    compileOnly("org.apache.spark:spark-streaming_${project.extra["scalaBinaryVersion"]}:${project.extra["sparkVersion"]}")
```

File: build.gradle.kts [hunk 7]
```diff
@@ -81,6 +85,11 @@ dependencies {
     testImplementation("org.mockito:mockito-junit-jupiter:${project.extra["mockitoVersion"]}")
     testImplementation("org.apiguardian:apiguardian-api:1.1.2") // https://github.com/gradle/gradle/issues/18627
 
+    testImplementation("org.apache.spark:spark-core_2.13:${project.extra["sparkVersion"]}")
```
Comment (cchantep) [inline build.gradle.kts:88]:
```suggestion
    testImplementation("org.apache.spark:spark-core_${project.extra["scalaBinaryVersion"]}:${project.extra["sparkVersion"]}")
```

File: build.gradle.kts [hunk 8]
```diff
@@ -81,6 +85,11 @@ dependencies {
...
+    testImplementation("org.apache.spark:spark-sql_2.13:${project.extra["sparkVersion"]}")
```
Comment (cchantep) [inline build.gradle.kts:89]:
```suggestion
    testImplementation("org.apache.spark:spark-sql_${project.extra["scalaBinaryVersion"]}:${project.extra["sparkVersion"]}")
```

File: build.gradle.kts [hunk 9]
```diff
@@ -81,6 +85,11 @@ dependencies {
...
+    testImplementation("org.apache.spark:spark-catalyst_2.13:${project.extra["sparkVersion"]}")
```
Comment (cchantep) [inline build.gradle.kts:90]:
```suggestion
    testImplementation("org.apache.spark:spark-catalyst_${project.extra["scalaBinaryVersion"]}:${project.extra["sparkVersion"]}")
```

File: build.gradle.kts [hunk 10]
```diff
@@ -81,6 +85,11 @@ dependencies {
...
+    testImplementation("org.apache.spark:spark-streaming_2.13:${project.extra["sparkVersion"]}")
```
Comment (cchantep) [inline build.gradle.kts:91]:
```suggestion
    testImplementation("org.apache.spark:spark-streaming_${project.extra["scalaBinaryVersion"]}:${project.extra["sparkVersion"]}")
```

Comment (dohr-michael) [reply build.gradle.kts]: I've integrated your suggestion + cross build with conditional sources in function of scala212 and greater.

Now need to add build of both version to have artifact properly added to nexus

**Comment by rozza** [comment]
Hi @dohr-michael & @cchantep

Thanks for the PR - I have added [SPARK-369](https://jira.mongodb.org/browse/SPARK-369) to track this work and hope to get it prioritized for the next main release.

**Comment by dohr-michael** [comment]
Hi @rozza 
If you need it, I can ask to our enterprise support (for mongodb) to 'vote' to this ticket.

**Comment by cchantep** [comment]
A support case is already created, waiting it to be escaladed.

**Comment by rozza** [comment]
Hi @dohr-michael & @cchantep,

Once again thanks for the PR and the feedback. I just wanted to let you know that this will be in the next release of the Spark connector and a SNAPSHOT build should be out later this week. I'll manually rebase and merge these PR commits and will update this ticket when its in main.

Ross

**Comment by rozza** [comment]
Rebased and merged in main :)

**Pull request closed without merging**